src/client/Microsoft.Identity.Client/Platforms/Android/Broker/AndroidContentProviderBroker.cs

```diff
@@ -0,0 +1,349 @@
+using Microsoft.Identity.Client.ApiConfig.Parameters;
+using Microsoft.Identity.Client.Core;
+using Microsoft.Identity.Client.Internal.Broker;
+using Microsoft.Identity.Client.Internal.Requests;
+using Microsoft.Identity.Client.OAuth2;
+using Microsoft.Identity.Client.UI;
+using AndroidNative = Android;
+using AndroidUri = Android.Net.Uri;
+
+namespace Microsoft.Identity.Client.Platforms.Android.Broker
+{
+    [AndroidNative.Runtime.Preserve(AllMembers = true)]
+    internal class AndroidContentProviderBroker : IBroker
+    {
+        private readonly AndroidBrokerHelper _brokerHelper;
+        private readonly ICoreLogger _logger;
+        private Activity _parentActivity;
+        private string _negotiatedBrokerProtocalKey = String.Empty;
+
+        public AndroidContentProviderBroker(CoreUIParent uiParent, ICoreLogger logger)
+        {
+            _parentActivity = uiParent?.Activity;
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            AuthenticationContinuationHelper.LastRequestLogger = _logger;
+            _brokerHelper = new AndroidBrokerHelper(Application.Context, logger);
+        }
+
+        public bool IsBrokerInstalledAndInvokable()
+        {
+            return _brokerHelper.IsBrokerInstalledAndInvokable();
+        }
+
+        public async void InitiateBrokerHandshakeAsync()
+        {
+            using (_logger.LogMethodDuration())
+            {
+                try
+                {
+                    Bundle HandshakeBundleResult = await GetHandshakeBundleResultFromBrokerAsync().ConfigureAwait(false);
+
+                    _negotiatedBrokerProtocalKey = _brokerHelper.GetProtocalKeyFromHandshakeResult(HandshakeBundleResult);
+                }
+                catch
+                {
+                    throw;
+                }
+            }
+        }
+
+        private async Task<Bundle> GetHandshakeBundleResultFromBrokerAsync()
+        {
+            var bundle = _brokerHelper.GetHandshakeOperationBundle();
+            var OperationBundleJSON = _brokerHelper.SearializeBundleToJSON(bundle);
+            return await PerformContentResolverOperationAsync(ContentResolverOperation.hello, OperationBundleJSON).ConfigureAwait(false);
+        }
+
+        public async Task<MsalTokenResponse> AcquireTokenInteractiveAsync(
+            AuthenticationRequestParameters authenticationRequestParameters,
+            AcquireTokenInteractiveParameters acquireTokenInteractiveParameters)
+        {
+            _brokerHelper.CheckPowerOptimizationStatus();
+
+            InitiateBrokerHandshakeAsync();
+
+            var brokerRequest = PrepareInteractiveBrokerRequest(authenticationRequestParameters, acquireTokenInteractiveParameters);
+
+            return await PerformAcquireTokenInteractiveAsync(brokerRequest).ConfigureAwait(false);
+        }
+
+        private BrokerRequest PrepareInteractiveBrokerRequest(AuthenticationRequestParameters authenticationRequestParameters,
```

Comment:
I would avoid these very small 1-2 lines of code methods. They fragment the code too much. Methods should also not be longer than ~50 lines or so (best description I know around this is `you should not have to scroll when reading the code of a method`)

Reply:
I agree with Bogdan on this.

---
In reply to: [569703588](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/2380#discussion_r569703588) [](ancestors = 569703588)